### PR TITLE
2D Polyline simplification: Also simplify overlapping vertex sequences

### LINKED ISSUE #5172: Polyline Simplification:  simplify overlapping sub-polylines
## Issue Details

A user on [stackoverflow](https://stackoverflow.com/questions/64779837/is-there-any-way-to-get-cgals-polyline-simplification-to-work-for-inner-shared) had polylines of several provinces of a country and only the country border got simplified. 
This was because the province borders where there twice.

We could put polylines into an adjacency list graph and calll `CGAL::split_graph_into_polylines()`  so that we had province borders only  once.  This is roughly coded [here](https://gist.github.com/afabri/87619dd7e32ae33494e375804c697bfa).

But it would even be better if the simplification algorithm could simplify overlapping polylines.

### PR DESCRIPTION
**Pull request by afabri**
## Summary of Changes

When two polylines where overlapping, that is having the same (potentially reversed) sub-sequence of vertices, they were considered as unremovable.      A posting on stackoverflow led to the feature request #5172 which gets implemented by this PR.

### Todo
- [x] Deal with polylines with sub-sequences `p-q-p`.  They were unremovable, and are not handled yet. Together with @sloriot we decided that such a sequence does not get simplified.
- [x] Write into changes.txt as what is considered unremovable has changed.  We consider this as a bug fix not as a new feature. 

## Release Management

* Affected package(s):  Polyline_simplification_2
* Issue(s) solved (if any): fix #5172
* Feature/Small Feature (if any):
* Link to compiled documentation (obligatory for small feature) [*wrong link name to be changed*](httpssss://wrong_URL_to_be_changed/Manual/Pkg)
* License and copyright ownership: not changed

### CODE REVIEW

**Review by sloriot — COMMENTED** [review]
We should have a meeting to discuss the changes, I'm not familiar enough with the code to give a deep review.

File: Polyline_simplification_2/include/CGAL/Polyline_simplification_2/simplify.h [hunk 1]
```diff
@@ -71,6 +75,11 @@ class Polyline_simplification_2
   StopFunction stop;
   std::size_t pct_initial_number_of_vertices, number_of_unremovable_vertices;
 
+  std::map<Vertex_handle, std::list<Vertices_in_constraint_iterator> > vertex_to_iterator;
```
[the remark this answers is not in the document]
Comment (afabri) [reply Polyline_simplification_2/include/CGAL/Polyline_simplification_2/simplify.h:78]: done

File: Polyline_simplification_2/include/CGAL/Polyline_simplification_2/simplify.h [hunk 2]
```diff
@@ -124,24 +148,67 @@ class Polyline_simplification_2
...
-        } else {
-          vertices.insert(*it);
-        }
-      }
+      it = pct.vertices_in_constraint_end(cid);
       it = boost::prior(it);
       (*it)->set_removable(false);
     }
+
+    std::map<Vertex_handle, int> degrees;
```
[the remark this answers is not in the document]
Comment (afabri) [reply Polyline_simplification_2/include/CGAL/Polyline_simplification_2/simplify.h:165]: done

File: Polyline_simplification_2/include/CGAL/Polyline_simplification_2/simplify.h [hunk 3]
```diff
@@ -124,24 +148,67 @@ class Polyline_simplification_2
     delete mpq;
   }
 
+  // endpoints of constraints are unremovable
+  // vertices which are not endpoint and have != 2 incident constained edges are unremovable
```
[the remark this answers is not in the document]
Comment (afabri) [reply Polyline_simplification_2/include/CGAL/Polyline_simplification_2/simplify.h:152]: fixed

File: Polyline_simplification_2/include/CGAL/Polyline_simplification_2/simplify.h [hunk 4]
```diff
@@ -239,10 +308,15 @@ class Polyline_simplification_2
...
   {
     int id = 0;
+    /*
     Constraint_iterator b = pct.constraints_begin(), e = pct.constraints_end();
     for(; b!=e; ++b){
       id = initialize_indices(*b, id);
     }
+    */
+    for(Finite_vertices_iterator it =  pct.finite_vertices_begin(); it != pct.finite_vertices_end(); ++it){
+      vertex_index_map[it] = id++;
```
[the remark this answers is not in the document]
Comment (afabri) [reply Polyline_simplification_2/include/CGAL/Polyline_simplification_2/simplify.h:318]: I didn't want to make the requirements more complex.

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.3-Ic-74.shtml

**Comment by adisidev** [comment]
This algorithm can get rather slow when one has upwards of a 1,000,000 points to simplify. Thus, would it by any chance be possible to run the algorithm in parallel somehow? Perhaps using OpenMP? From my naive look at the implementation, nothing seems to strongly indicate that the simplification would have to run strictly serially.

**Pull request merged**